Add appendix past changes and interoperability considerations (#171)
ietf-wg-cellar/flac-specification: rfc_backmatter.md

## Rice coding
When folding (i.e. zig-zag encoding) the residual sample values, no extra bits are needed when the absolute value of each residual sample is first stored in an unsigned data type of the size of the last step, then doubled and then has one subtracted depending on whether the residual sample was positive or negative. Many implementations however choose to require one extra bit of data type size so zig-zag encoding can happen in one step and without a cast instead of the procedure described in the previous sentence.

# Past format changes

> This informational appendix documents what changes were made to the FLAC format over the years. This information might be of use when encountering FLAC files that were made with software following the format as it was specified initially.

Author: Would standardization be an improvement over stabilization?

I don't really like 'in earlier days', it feels a little vague. What about `This information might be of use when encountering FLAC files that were made with software following the format as it was before the changes documented in this appendix.`

> This informational appendix documents what changes were made to the FLAC format over the years. This information might be of use when encountering FLAC files that were made with software following the format as it was specified initially.

Author: ```suggestion
This informational appendix documents what changes were made to the FLAC format over the years. This information might be of use when encountering FLAC files that were made with software following the format as it was before the changes documented in this appendix.
```

The FLAC format was first specified in December 2000 and the bitstream format was considered frozen with the release of FLAC 1.0 (the reference encoder/decoder) in July 2001. Only changes made since this first stable release are considered in this appendix. Changes made to the FLAC subset definition are not considered.

## Addition of blocksize strategy flag

> Perhaps the largest backwards incompatible change to the specification was published in July 2007. Before this change, variable blocksize streams were not explicitly marked as such by a flag bit in the frame header. A decoder had two ways to detect a variable blocksize stream, either by comparing the minimum and maximum blocksize in the STREAMINFO metadata block (which are equal in case of a fixed blocksize stream), or, in case a STREAMINFO metadata block was not present in a file, by detecting a change of blocksize during a stream, which could in theory not happen at all. As the meaning of the coded number number in the frame header depends on whether or not a stream is variable blocksize, this presented a problem: the meaning of the coded number could not be reliably determined. To fix this problem, one of the reserved bits was changed to be used as a blocksize strategy flag. [See also the section frame header](#frame-header).

Author: This addition would contradict the specification, because all streams must have a streaminfo block. However, on multicasting, it is possible the streaminfo block was not received. Perhaps I should rephrase accordingly.

`in case a STREAMINFO metadata block was not present in a file` -> `in case a decoder did not receive a STREAMINFO metadata block` but this might be confusing. Perhaps `in case a FLAC stream is multicast and a decoder did not receive a STREAMINFO metadata block`

> Along with the addition of a new flag, the meaning of the [blocksize bits](#blocksize-bits) was subtly changed. Initially, blocksize bits 0b0001-0b0101 and 0b1000-0b1111 could only be used for fixed blocksize streams, while 0b0110 and 0b0111 could be used for both fixed blocksize and variable blocksize streams. With the change these restrictions were lifted and 0b0001-0b1111 are now used for both variable blocksize and fixed blocksize streams.

Reviewer: > It would perhaps be better to recommend not to use variable blocksize streams at all if wide interoperability is a concern. That doesn't fit in this part though, so I think this concern (and the others suggested) are better moved to a wiki page and then linked from here and the section implementation status.

IMO and if I remember well IETF docs have sometimes a paragraph about interoperability.

It may be good to add another appendix at the end of this patch about interoperability issues e.g.

> # Wide interoperability
>
> At the time of writing of this document, several hardware FLAC decoders are already in the wild and can not be easily upgraded. Some of them don't support variable blocksize streams. Encoders willing to maintain interoperability with a wide range of hardware decoders are RECOMMENDED to avoid variable blocksize.`

## Restriction of encoded residual samples

Another change to the specification was deemed necessary during standardization by the CELLAR working group of the IETF. As specified in [section coded residual](#coded-residual) a limit is imposed on residual samples. This limit was not specified prior to the IETF standardization effort. However, as far as was known to the working group, no FLAC encoder at that time produced FLAC files containing residual samples exceeding this limit. This is mostly because it is very unlikely to encounter residual samples exceeding this limit when encoding 24-bit PCM, and encoding of PCM with higher bitdepths was not yet implemented in any known encoder. In fact, these FLAC encoders would produce corrupt files upon being triggered to produce such residual samples and it is unlikely any non-experimental encoder would ever do so, even when presented with crafted material. Therefore, it was not expected existing implementation would be rendered non-compliant by this change.

## Addition of 5-bit Rice parameter

> One significant addition to the format was the residual coding method using a 5-bit Rice parameter. Prior to publication of this addition in July 2007, there was only one residual coding method specified, a partitioned Rice code with a 4-bit Rice parameter. The range offered by this proved too small when encoding 24-bit PCM, therefore a second residual coding method was specified identical to the first but with a 5-bit Rice parameter.

Reviewer: \+ "Encoders willing to avoid interoperability issues with very old decoders SHOULD NOT use 5-bit Rice parameter." if the advice make sense.

Author: > @ktmf01 is it also possible to have files in the test files which have the newest behaviors, especially the 5-bit Rice coding? So we could test explicitly if a decoder is very old or not.

It is not tested separately, but almost all 24-bit files use 5-bit rice parameters somewhere in the file. Testbench files 28 - 35, file 44 and files 62 and 63 use these.

The reference FLAC encoder never uses 5-bit rice parameters for 16-bit files for compatibility reasons, but ffmpeg ensures no such thing.

For the variable blocksize comment, file 27 in the testbench is specifically to test new decoders being able to decode according to the old format (many are unable to do so). Files 24-26 follow current spec.

> + "Encoders willing to avoid interoperability issues with very old decoders SHOULD NOT use 5-bit Rice parameter." if the advice make sense.

This could hurt compression for 24-bit files rather badly. Something like this perhaps?

> Encoders willing to maintain interoperability with very old decoders are RECOMMENDED to avoid coding 5-bit Rice parameters when encoding PCM with a bit depth not exceeding 16. Avoiding the use of 5-bit Rice parameters for all bit depths could significantly reduce compression.

Reviewer: > The reference FLAC encoder never uses 5-bit rice parameters for 16-bit files for compatibility reasons, but ffmpeg ensures no such thing.

Good to know. I'll check if it is easy to have an option for forcing 4-bit Rice for 16-bit in FFmpeg.

> This could hurt compression for 24-bit files rather badly. Something like this perhaps?
>
> > Encoders willing to maintain interoperability with very old decoders are RECOMMENDED to avoid coding 5-bit Rice parameters when encoding PCM with a bit depth not exceeding 16. Avoiding the use of 5-bit Rice parameters for all bit depths could significantly reduce compression.

Agreed.

Could be moved to Wide interoperability appendix if you agree to have this appendix.
